File: 2020/day07.js
const fs = require('fs');

fs.readFile('input/day07', 'utf8', (_, data) => {
  const colors = data.trim().split('\n').reduce((acc, line) => ({
    ...acc,
    ...parseLine(line)
  }));

  const countBagsIn = (color, factor = 1) => Object
    .entries(colors[color])
    .reduce((count, [subColor, qty]) => (
      count + factor * qty + countBagsIn(subColor, factor * qty)
    ), 0);

  console.log('Part 1:', filterBags(colors).length);
  console.log('Part 2:', countBagsIn('shiny gold'));
});

const parseLine = line => {
  const [_, color, contents] = line.match(/^(\S+ \S+) bags contain (.*)\.$/);

  return {
    [color]: contents.split(', ').reduce((acc, content) => {
      const match = content.match(/^(\d+) (\S+ \S+) bags?$/);
      return match ? { ...acc, [match[2]]: match[1] } : acc;
    }, {}),
  };
};

const filterBags = (colors, bagsAcc = []) => {
  const [newBagsAcc, keepGoing] = Object
    .entries(colors)
    .reduce(colorsReducer, [bagsAcc, false]);

  return keepGoing ? filterBags(colors, newBagsAcc) : newBagsAcc;
};

const colorsReducer = ([bagsAcc, keepGoing], [color, subColors]) => Object
  .keys(subColors)
  .reduce(subColorsReducer(color, bagsAcc), [bagsAcc, keepGoing]);

const subColorsReducer = (color, bagsAcc) => ([acc, keep], k) => (
  (k === 'shiny gold' || bagsAcc.includes(k)) && !acc.includes(color)
    ? [[...acc, color], true]
    : [acc, keep]
);

File: 2020/day07.hs
import Data.Char (isDigit)
import Data.List.Split (splitOn)
import Data.Set as S (Set, empty, insert, member)
import Data.Map.Strict as M (Map, (!), empty, insert, member, toList)

type Bags = Map String (Map String Int)

main :: IO ()
main = do
  bags <- foldl parseLines M.empty . lines <$> readFile "input/day07"

  let bags' = filter (canContainShinyGold bags S.empty . fst) $ toList bags
      bagsInShinyGold = countBagsIn bags "shiny gold" 1

  putStrLn $ "Part 1: " ++ show (length bags')
  putStrLn $ "Part 2: " ++ show bagsInShinyGold

parseLines :: Bags -> String -> Bags
parseLines bags line = M.insert colorName (parseContents contents) bags
  where
    contents = splitOn ", " contentString
    [colorName, contentString] = splitOn " bags contain " line
    parseContents (content : rest)
      | isDigit (head content) =
        let (qty : subColorWords) = take 3 $ words content
         in M.insert (unwords subColorWords) (read qty) $ parseContents rest
      | otherwise = parseContents rest
    parseContents [] = M.empty

canContainShinyGold :: Bags -> Set String -> String -> Bool
canContainShinyGold bags visited color
  | color `S.member` visited = False
  | "shiny gold" `M.member` subColors = True
  | otherwise = any (canContainShinyGold bags (S.insert color visited) . fst)
              $ toList subColors
  where
    subColors = bags ! color

countBagsIn :: Bags -> String -> Int -> Int
countBagsIn bags color factor =
  foldl countInSubColors 0 . M.toList $ bags ! color
  where
    countInSubColors acc (sub, qty) =
      acc + factor * qty + countBagsIn bags sub (factor * qty)
